kube/src/api/admission.rs

```diff
@@ -0,0 +1,280 @@
+    pub patch: Option<Vec<u8>>,
+    /// The type of Patch. Currently we only allow "JSONPatch".
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub patch_type: Option<PatchType>,
+    /// An unstructured key value map set by remote admission controller (e.g.
+    /// error=image-blacklisted). MutatingAdmissionWebhook and
+    /// ValidatingAdmissionWebhook admission controller will prefix the keys with
+    /// admission webhook name (e.g.
+    /// imagepolicy.example.com/error=image-blacklisted). AuditAnnotations will
+    /// be provided by the admission webhook to add additional context to the
+    /// audit log for this request.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub audit_annotations: Option<HashMap<String, String>>,
+    /// A list of warning messages to return to the requesting API client.
+    /// Warning messages describe a problem the client making the API request
+    /// should correct or be aware of. Limit warnings to 120 characters if
+    /// possible. Warnings over 256 characters and large numbers of warnings may
+    /// be truncated.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub warnings: Option<Vec<String>>,
+}
+
+impl AdmissionResponse {
+    /// Constructs an invalid [`AdmissionResponse`]. It doesn't copy the uid from
+    /// the corresponding [`AdmissionRequest`], so should only be used when the
+    /// original request cannot be read.
+    pub fn invalid<T: ToString>(reason: T) -> Self {
+        Self {
+            uid: Default::default(),
+            allowed: false,
+            result: Status {
+                reason: Some(reason.to_string()),
+                ..Default::default()
+            },
+            patch: None,
+            patch_type: None,
+            audit_annotations: None,
+            warnings: None,
+        }
+    }
+
+    /// Constructs an allowed [`AdmissionResponse`] from an [`AdmissionRequest`].
+    pub fn from_request<T: Resource>(req: &AdmissionRequest<T>) -> Self {
+        Self {
+            uid: req.uid.clone(),
+            allowed: true,
+            result: Default::default(),
+            patch: None,
+            patch_type: None,
+            audit_annotations: None,
+            warnings: None,
+        }
+    }
+
+    /// Deny the request with a reason. The reason will be sent to the original
+    /// caller.
+    pub fn deny<T: ToString>(mut self, reason: T) -> Self {
+        self.allowed = false;
+        self.result.message = Some(reason.to_string());
+
+        self
+    }
+
+    /// Add JSON patches to the response, modifying the object from the request.
+    pub fn with_patch(mut self, patch: json_patch::Patch) -> Self {
+        self.patch = match serde_json::to_vec(&patch) {
+            Ok(v) => {
+                self.patch_type = Some(PatchType::JsonPatch);
+                Some(v)
+            }
+            Err(err) => {
+                self = self.deny(err.to_string());
+                None
```

Comment:
Guess this is if the user makes an invalid json patch?

Maybe this should be propagated as an error type instead (Result<Self>) rather than the admission controller just rejecting every object. Technically if people go to debug the controller here, there would be nothing in the logs.